GPUP deadlock shutting down RemoteRenderingBackend while drawing a MediaPlayer

kkinnunen-apple opened the pull request:
#### 63ca1cca5ba566ac40e5786711aab59493f6cbb1
<pre>
GPUP deadlock shutting down RemoteRenderingBackend while drawing a MediaPlayer
<a href="https://bugs.webkit.org/show_bug.cgi?id=243838">https://bugs.webkit.org/show_bug.cgi?id=243838</a>
<a href="https://rdar.apple.com/98523368">rdar://98523368</a>

Reviewed by Youenn Fablet.

GPUP would deadlock in scenario where:
 - RemoteRenderingBackend would be torn down due to, say, connection
   closing. GPUP main thread would wait to complete to shut down RRB
   work queue.
 - RRB would execute RemoteDisplayListRecorder::paintFrameForMedia
   which would try to run a synchronous task from RRB work queue
   in main thread, painting the media player to a ImageBuffer
Since main thread would be waiting, it wouldn&apos;t run the
paintFrameForMedia.

The paintFrameForMedia is a command that draws current frame of media player
into the GraphicsContext. This doesn&apos;t particularly make sense, as
the &quot;current time&quot; would refer to the original draw time in WCP, but
as implemented, the frame would be snapshot at the actual draw command
execution time.

Instead, do:
 - Obtain reference to current video frame from the media player
 - Draw that video frame

In GPUP, remote media players create the frames in their queue
(main thread). Remote rendering accesses the video frames by their
read references through the video frame object heap.

* Source/WebCore/html/HTMLVideoElement.cpp:
(WebCore::HTMLVideoElement::paint):
Added the plain paint() function to signify paint from render objects.
This probably existed at some point, as a caller to VideoPlayer::paint().
Otherwise snapshots and WKTR captures would make the video element
signal that the video element is needed for 2d context/WebGL, and
force updates, failing media/video-page-visibility-restriction.html.
The added function uses MediaPlayer::paint(), which was dead code
previously.

(WebCore::HTMLVideoElement::paintCurrentFrameInContext):
* Source/WebCore/html/HTMLVideoElement.h:
* Source/WebCore/html/ImageBitmap.cpp:
(WebCore::ImageBitmap::createCompletionHandler):
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
(WebCore::CanvasRenderingContext2DBase::drawImage):
* Source/WebCore/platform/VideoFrame.cpp:
(WebCore::VideoFrame::draw):
(WebCore::VideoFrame::paintInContext): Deleted.
* Source/WebCore/platform/VideoFrame.h:
* Source/WebCore/platform/graphics/BifurcatedGraphicsContext.cpp:
(WebCore::BifurcatedGraphicsContext::drawVideoFrame):
(WebCore::BifurcatedGraphicsContext::paintFrameForMedia): Deleted.
* Source/WebCore/platform/graphics/BifurcatedGraphicsContext.h:
* Source/WebCore/platform/graphics/GraphicsContext.cpp:
(WebCore::GraphicsContext::drawVideoFrame):
(WebCore::GraphicsContext::paintFrameForMedia): Deleted.
(WebCore::GraphicsContext::paintVideoFrame): Deleted.
* Source/WebCore/platform/graphics/GraphicsContext.h:
* Source/WebCore/platform/graphics/GraphicsContextGL.cpp:
(WebCore::GraphicsContextGL::videoFrameToImage):
* Source/WebCore/platform/graphics/NullGraphicsContext.h:
* Source/WebCore/platform/graphics/cv/VideoFrameCV.mm:
(WebCore::VideoFrame::draw):
(WebCore::VideoFrame::paintInContext): Deleted.
* Source/WebCore/platform/graphics/displaylists/DisplayListItem.h:
* Source/WebCore/platform/graphics/displaylists/DisplayListItems.cpp:
(WebCore::DisplayList::PaintFrameForMedia::PaintFrameForMedia): Deleted.
(WebCore::DisplayList::PaintFrameForMedia::apply const): Deleted.
(WebCore::DisplayList::PaintFrameForMedia::dump const): Deleted.
* Source/WebCore/platform/graphics/displaylists/DisplayListItems.h:
(WebCore::DisplayList::PaintFrameForMedia::identifier const): Deleted.
(WebCore::DisplayList::PaintFrameForMedia::destination const): Deleted.
(WebCore::DisplayList::PaintFrameForMedia::isValid const): Deleted.
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.cpp:
(WebCore::DisplayList::Recorder::drawVideoFrame):
(WebCore::DisplayList::Recorder::paintFrameForMedia): Deleted.
(WebCore::DisplayList::Recorder::paintVideoFrame): Deleted.
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.h:
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorderImpl.cpp:
(WebCore::DisplayList::RecorderImpl::recordDrawVideoFrame):
(WebCore::DisplayList::RecorderImpl::recordPaintFrameForMedia): Deleted.
(WebCore::DisplayList::RecorderImpl::recordPaintVideoFrame): Deleted.
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorderImpl.h:
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::paint):
* Source/WebCore/platform/graphics/gstreamer/VideoFrameGStreamer.cpp:
(WebCore::VideoFrame::draw):
(WebCore::VideoFrame::paintInContext): Deleted.
* Source/WebCore/platform/graphics/nicosia/cairo/NicosiaCairoOperationRecorder.cpp:
(Nicosia::CairoOperationRecorder::drawVideoFrame):
(Nicosia::CairoOperationRecorder::paintFrameForMedia): Deleted.
* Source/WebCore/platform/graphics/nicosia/cairo/NicosiaCairoOperationRecorder.h:
* Source/WebCore/rendering/RenderVideo.cpp:
(WebCore::RenderVideo::paintReplaced):
* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.cpp:
(WebKit::RemoteDisplayListRecorder::drawVideoFrame):
(WebKit::RemoteDisplayListRecorder::paintFrameForMedia): Deleted.
(WebKit::RemoteDisplayListRecorder::paintVideoFrame): Deleted.
* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.h:
* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.h:
* Source/WebKit/Shared/DisplayListArgumentCoders.serialization.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/GPU/graphics/RemoteDisplayListRecorderProxy.cpp:
(WebKit::RemoteDisplayListRecorderProxy::recordDrawVideoFrame):
(WebKit::RemoteDisplayListRecorderProxy::recordPaintFrameForMedia): Deleted.
(WebKit::RemoteDisplayListRecorderProxy::recordPaintVideoFrame): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/RemoteDisplayListRecorderProxy.h:

Canonical link: <a href="https://commits.webkit.org/283637@main">https://commits.webkit.org/283637@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/47b25a9fd5dc27e6fac41dbe82dc5b642834f253

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/66883 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/46258 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/19505 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/70918 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/18016 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/69001 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/54057 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/17796 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/53649 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/12124 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/69950 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/42558 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/57870 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/34193 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/39229 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/15262 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/16370 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/61145 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/15603 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/72619 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/10840 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/14950 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/61061 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/10872 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/57927 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/61217 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14774 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/8904 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/2522 "Passed tests") | | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/42065 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/43142 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/44325 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/42885 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr reviewed (commented):
Nice fix!

youennf commented on Source/WebCore/html/HTMLVideoElement.h:
```diff
@@ -78,6 +78,9 @@ class HTMLVideoElement final : public HTMLMediaElement, public Supplementable<HT
 
     RefPtr<ImageBuffer> createBufferForPainting(const FloatSize&, RenderingMode, const DestinationColorSpace&, ImageBufferPixelFormat) const;
 
+    // Used by render painting. Best effort, ionly paint if we already have an image generator or video output available.
```
```suggestion
    // Used by render painting. Best effort, only paint if we already have an image generator or video output available.
```

youennf commented on Source/WebCore/html/HTMLVideoElement.h:
```diff
@@ -78,6 +78,9 @@ class HTMLVideoElement final : public HTMLMediaElement, public Supplementable<HT
     RefPtr<ImageBuffer> createBufferForPainting(const FloatSize&, RenderingMode, const DestinationColorSpace&, ImageBufferPixelFormat) const;
 
+    // Used by render painting. Best effort, ionly paint if we already have an image generator or video output available.
+    void paint(GraphicsContext&, const FloatRect&);
```
It is a preexisting issue but It would be nice to disambiguate paint and paintCurrentFrameInContext.
Maybe paint could be renamed to paintForRendering or draw?

youennf commented on Source/WebCore/html/HTMLVideoElement.cpp:
```diff
@@ -345,6 +346,14 @@ RefPtr<ImageBuffer> HTMLVideoElement::createBufferForPainting(const FloatSize& s
+    RefPtr player = this->player();
+    if (!player)
+        return;
+    player->playerPrivate()->paint(context, destRect);
```
```suggestion
    player->paint(context, destRect);
```

youennf replied:
Also, if we do this, we probably end up in MediaPlayerPrivateRemote::paint which triggers two sync IPC, one to get the video frame, and another one to convert it to native video frame?
Is there a way to remove the conversion and call context. drawVideoFrame instead of context.drawNativeImage?

kkinnunen-apple replied:
> Is there a way to remove the conversion and call context. drawVideoFrame instead of context.drawNativeImage?

Yes, good catch. That was the point of the patch all along, I just assumed it was already working like this but of it wasn't.  Updated..

youennf commented on Source/WebCore/html/HTMLVideoElement.cpp:
```diff
@@ -355,7 +364,7 @@ void HTMLVideoElement::paintCurrentFrameInContext(GraphicsContext& context, cons
         visibilityStateChanged();
     }
-    context.paintFrameForMedia(*player, destRect);
+    player->playerPrivate()->paintCurrentFrameInContext(context, destRect);
```
It would not hurt to introduce a MediaPlayer:: paintCurrentFrameInContext.
Ditto for the double sync IPC, right?

youennf reviewed (commented):
This sounds ok to me, some questions though below.

youennf reviewed (approved):
LGTM

kkinnunen-apple committed:
GPUP deadlock shutting down RemoteRenderingBackend while drawing a MediaPlayer
https://bugs.webkit.org/show_bug.cgi?id=243838
rdar://98523368

Reviewed by Youenn Fablet.

GPUP would deadlock in scenario where:
 - RemoteRenderingBackend would be torn down due to, say, connection
   closing. GPUP main thread would wait to complete to shut down RRB
   work queue.
 - RRB would execute RemoteDisplayListRecorder::paintFrameForMedia
   which would try to run a synchronous task from RRB work queue
   in main thread, painting the media player to a ImageBuffer
Since main thread would be waiting, it wouldn't run the
paintFrameForMedia.

The paintFrameForMedia is a command that draws current frame of media player
into the GraphicsContext. This doesn't particularly make sense, as
the "current time" would refer to the original draw time in WCP, but
as implemented, the frame would be snapshot at the actual draw command
execution time.

Instead, do:
 - Obtain reference to current video frame from the media player
 - Draw that video frame

In GPUP, remote media players create the frames in their queue
(main thread). Remote rendering accesses the video frames by their
read references through the video frame object heap.

* Source/WebCore/html/HTMLVideoElement.cpp:
(WebCore::HTMLVideoElement::paint):
Added the plain paint() function to signify paint from render objects.
This probably existed at some point, as a caller to VideoPlayer::paint().
Otherwise snapshots and WKTR captures would make the video element
signal that the video element is needed for 2d context/WebGL, and
force updates, failing media/video-page-visibility-restriction.html.
The added function uses MediaPlayer::paint(), which was dead code
previously.

(WebCore::HTMLVideoElement::paintCurrentFrameInContext):
* Source/WebCore/html/HTMLVideoElement.h:
* Source/WebCore/html/ImageBitmap.cpp:
(WebCore::ImageBitmap::createCompletionHandler):
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
(WebCore::CanvasRenderingContext2DBase::drawImage):
* Source/WebCore/platform/VideoFrame.cpp:
(WebCore::VideoFrame::draw):
(WebCore::VideoFrame::paintInContext): Deleted.
* Source/WebCore/platform/VideoFrame.h:
* Source/WebCore/platform/graphics/BifurcatedGraphicsContext.cpp:
(WebCore::BifurcatedGraphicsContext::drawVideoFrame):
(WebCore::BifurcatedGraphicsContext::paintFrameForMedia): Deleted.
* Source/WebCore/platform/graphics/BifurcatedGraphicsContext.h:
* Source/WebCore/platform/graphics/GraphicsContext.cpp:
(WebCore::GraphicsContext::drawVideoFrame):
(WebCore::GraphicsContext::paintFrameForMedia): Deleted.
(WebCore::GraphicsContext::paintVideoFrame): Deleted.
* Source/WebCore/platform/graphics/GraphicsContext.h:
* Source/WebCore/platform/graphics/GraphicsContextGL.cpp:
(WebCore::GraphicsContextGL::videoFrameToImage):
* Source/WebCore/platform/graphics/NullGraphicsContext.h:
* Source/WebCore/platform/graphics/cv/VideoFrameCV.mm:
(WebCore::VideoFrame::draw):
(WebCore::VideoFrame::paintInContext): Deleted.
* Source/WebCore/platform/graphics/displaylists/DisplayListItem.h:
* Source/WebCore/platform/graphics/displaylists/DisplayListItems.cpp:
(WebCore::DisplayList::PaintFrameForMedia::PaintFrameForMedia): Deleted.
(WebCore::DisplayList::PaintFrameForMedia::apply const): Deleted.
(WebCore::DisplayList::PaintFrameForMedia::dump const): Deleted.
* Source/WebCore/platform/graphics/displaylists/DisplayListItems.h:
(WebCore::DisplayList::PaintFrameForMedia::identifier const): Deleted.
(WebCore::DisplayList::PaintFrameForMedia::destination const): Deleted.
(WebCore::DisplayList::PaintFrameForMedia::isValid const): Deleted.
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.cpp:
(WebCore::DisplayList::Recorder::drawVideoFrame):
(WebCore::DisplayList::Recorder::paintFrameForMedia): Deleted.
(WebCore::DisplayList::Recorder::paintVideoFrame): Deleted.
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.h:
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorderImpl.cpp:
(WebCore::DisplayList::RecorderImpl::recordDrawVideoFrame):
(WebCore::DisplayList::RecorderImpl::recordPaintFrameForMedia): Deleted.
(WebCore::DisplayList::RecorderImpl::recordPaintVideoFrame): Deleted.
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorderImpl.h:
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::paint):
* Source/WebCore/platform/graphics/gstreamer/VideoFrameGStreamer.cpp:
(WebCore::VideoFrame::draw):
(WebCore::VideoFrame::paintInContext): Deleted.
* Source/WebCore/platform/graphics/nicosia/cairo/NicosiaCairoOperationRecorder.cpp:
(Nicosia::CairoOperationRecorder::drawVideoFrame):
(Nicosia::CairoOperationRecorder::paintFrameForMedia): Deleted.
* Source/WebCore/platform/graphics/nicosia/cairo/NicosiaCairoOperationRecorder.h:
* Source/WebCore/rendering/RenderVideo.cpp:
(WebCore::RenderVideo::paintReplaced):
* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.cpp:
(WebKit::RemoteDisplayListRecorder::drawVideoFrame):
(WebKit::RemoteDisplayListRecorder::paintFrameForMedia): Deleted.
(WebKit::RemoteDisplayListRecorder::paintVideoFrame): Deleted.
* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.h:
* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.h:
* Source/WebKit/Shared/DisplayListArgumentCoders.serialization.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/GPU/graphics/RemoteDisplayListRecorderProxy.cpp:
(WebKit::RemoteDisplayListRecorderProxy::recordDrawVideoFrame):
(WebKit::RemoteDisplayListRecorderProxy::recordPaintFrameForMedia): Deleted.
(WebKit::RemoteDisplayListRecorderProxy::recordPaintVideoFrame): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/RemoteDisplayListRecorderProxy.h:

Canonical link: https://commits.webkit.org/283637@main
